File: apps/server/src/infra/yuqueClient.ts
/**
 * Yuque API Client — 对标 MetaBlog yuque.ts
 *
 * 同时支持语雀官方 Open API v2 和内部 Web API（Cookie 认证）。
 */

import type { PrismaClient } from "@prisma/client";
import type { AppConfig } from "./config.js";
import { getCredentialValue } from "./credentialVault.js";

const YUQUE_BASE = "https://www.yuque.com";

interface YuqueCredentials {
  session: string;
  ctoken: string;
}

export async function getYuqueCredentials(
  prisma: PrismaClient | undefined,
  config: AppConfig,
): Promise<YuqueCredentials> {
  let session = config.integrations.yuque.session || "";
  let ctoken = config.integrations.yuque.ctoken || "";

  if (prisma && (!session || !ctoken)) {
    session = (await getCredentialValue(prisma, "yuque", "yuque_session")) || session;
    ctoken = (await getCredentialValue(prisma, "yuque", "yuque_ctoken")) || ctoken;
  }

  if (!session) {
    throw new Error(
      "【语雀认证失败】未配置 YUQUE_SESSION。\n" +
        "1. 登录语雀网页版 https://www.yuque.com\n" +
        "2. F12 → Application → Cookies → https://www.yuque.com\n" +
        "3. 复制 _yuque_session 和 _ctoken 到 .env 或 Credentials 表（scope=yuque）",
    );
  }
  return { session, ctoken };
}

function buildHeaders(ctoken: string, referer?: string): Record<string, string> {
  const headers: Record<string, string> = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    Accept: "application/json, text/plain, */*",
    "X-CSRF-Token": ctoken,
    "X-Requested-With": "XMLHttpRequest",
  };
  if (referer) headers["Referer"] = referer;
  return headers;
}

export async function yuqueApi(
  method: string,
  path: string,
  options: {
    body?: unknown;
    query?: Record<string, string>;
    referer?: string;
    credentials?: YuqueCredentials;
  } = {},
): Promise<any> {
  const { session, ctoken } = options.credentials!;

  let url = `${YUQUE_BASE}${path.startsWith("/") ? path : `/${path}`}`;
  if (options.query) {
    const params = new URLSearchParams(options.query);
    url += "?" + params.toString();
  }

  const headers = buildHeaders(ctoken, options.referer);
  headers["Cookie"] = `_yuque_session=${session}; _ctoken=${ctoken}`;
  if (options.body && method !== "GET" && method !== "DELETE") {
    headers["Content-Type"] = "application/json";
  }

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 15000);

  try {
    const res = await fetch(url, {
      method,
      headers,
      body: options.body ? JSON.stringify(options.body) : undefined,
      signal: controller.signal,
    });
    return await res.json();
  } finally {
    clearTimeout(timeoutId);
  }
}

export async function yuqueOpenApi<T = unknown>(
  method: string,
  endpoint: string,
  options: { body?: unknown; query?: Record<string, string>; token?: string } = {},
): Promise<T> {
  const url = new URL(`https://www.yuque.com/api/v2${endpoint.startsWith("/") ? endpoint : `/${endpoint}`}`);
  if (options.query) {
    for (const [k, v] of Object.entries(options.query)) url.searchParams.set(k, v);
  }
  const headers: Record<string, string> = {
    "User-Agent": "KnowPilot/1.0",
    Accept: "application/json",
    "X-Auth-Token": options.token || "",
  };
  if (options.body && method !== "GET") {
    headers["Content-Type"] = "application/json";
  }
  const res = await fetch(url.toString(), {
    method,
    headers,
    body: options.body ? JSON.stringify(options.body) : undefined,
  });
  const data = (await res.json()) as any;
  if (!res.ok || data.status !== undefined) {
    throw new Error(`语雀 API 失败: ${data.message || res.statusText} (${res.status})`);
  }
  return data.data as T;
}

/* ─── 内部 Web API 快捷 ─── */

export async function yuqueListBooks(credentials: YuqueCredentials) {
  return yuqueApi("GET", "/api/books", { credentials });
}

export async function yuqueGetBookToc(bookId: string, credentials: YuqueCredentials) {
  return yuqueApi("GET", `/api/books/${bookId}/toc`, { credentials });
}

export async function yuqueGetDocWeb(docSlug: string, bookId: string, credentials: YuqueCredentials) {
  return yuqueApi("GET", `/api/docs/${docSlug}`, { query: { book_id: bookId }, credentials });
}

export async function yuqueCreateDoc(
  bookId: string,
  title: string,
  body: string,
  credentials: YuqueCredentials,
) {
  return yuqueApi(
    "POST",
    "/api/docs",
    {
      body: { book_id: bookId, title, body, format: "markdown" },
      referer: `${YUQUE_BASE}/api/docs`,
      credentials,
    },
  );
}

export async function yuqueUpdateDoc(
  docId: string,
  bookId: string,
  title: string,
  body: string,
  credentials: YuqueCredentials,
) {
  return yuqueApi(
    "PUT",
    `/api/docs/${docId}`,
    {
      body: { title, body, format: "markdown" },
      referer: `${YUQUE_BASE}/api/docs`,
      credentials,
    },
  );
}

export async function yuqueDeleteDoc(docId: string, bookId: string, credentials: YuqueCredentials) {
  return yuqueApi("DELETE", `/api/docs/${docId}`, {
    query: { book_id: bookId },
    referer: `${YUQUE_BASE}/api/docs`,
    credentials,
  });
}

/* ─── Open API v2 快捷 ─── */

export async function yuqueListRepos(token: string) {
  return yuqueOpenApi("GET", "/users/me/repos", { token });
}

export async function yuqueListDocs(namespace: string, token: string) {
  return yuqueOpenApi("GET", `/repos/${namespace}/docs`, { token });
}

export async function yuqueGetDocV2(namespace: string, slug: string, token: string) {
  if (!token) throw new Error("未配置语雀凭证（YUQUE_CTOKEN 或 Credential scope=yuque yuque_ctoken）");
  return yuqueOpenApi("GET", `/repos/${namespace}/docs/${slug}`, { token });
}

export async function yuqueCreateDocV2(
  namespace: string,
  title: string,
  body: string,
  token: string,
) {
  return yuqueOpenApi("POST", `/repos/${namespace}/docs`, {
    token,
    body: { title, body, public: 0 },
  });
}

export async function yuqueUpdateDocV2(
  namespace: string,
  slug: string,
  title: string,
  body: string,
  token: string,
) {
  return yuqueOpenApi("PUT", `/repos/${namespace}/docs/${slug}`, {
    token,
    body: { title, body },
  });
}

export async function yuqueDeleteDocV2(namespace: string, slug: string, token: string) {
  return yuqueOpenApi("DELETE", `/repos/${namespace}/docs/${slug}`, { token });
}

File: experiments/yuque-api/yuque_client.py
#!/usr/bin/env python3
"""
语雀 (Yuque) Web API Python 客户端（Cookie 认证）
来源：MetaBlog project/experiments/yuque-api/99_yuque_api_showcase.ipynb
"""

import os
import json
import time
import requests
from datetime import datetime
from pathlib import Path
from dotenv import load_dotenv

# 加载项目根目录 .env
load_dotenv(Path(__file__).resolve().parent.parent.parent / ".env")


class YuqueClient:
    """语雀 Web API 客户端(Cookie 认证)"""

    BASE_URL = "https://www.yuque.com"

    def __init__(self, session=None, ctoken=None):
        self.session = session or os.environ.get("YUQUE_SESSION")
        self.ctoken = ctoken or os.environ.get("YUQUE_CTOKEN")
        if not self.session:
            raise ValueError("请提供 _yuque_session Cookie 或设置 YUQUE_SESSION 环境变量")
        self.cookies = {
            "_yuque_session": self.session,
            "ctoken": self.ctoken or ""
        }
        self.headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            "Referer": "https://www.yuque.com",
            "Content-Type": "application/json"
        }
        self.session_req = requests.Session()
        self.session_req.cookies.update(self.cookies)
        self.session_req.headers.update(self.headers)
        self._doc_cache = {}  # 缓存 doc_id -> {slug, book_id}

    def request(self, method, endpoint, **kwargs):
        """发送 HTTP 请求"""
        url = f"{self.BASE_URL}{endpoint}"
        resp = self.session_req.request(method, url, **kwargs)
        resp.raise_for_status()
        return resp.json()

    def health_check(self):
        """健康检查：验证 Cookie 是否有效"""
        try:
            result = self.request("GET", "/api/mine")
            user = result.get("data", {})
            return {
                "ok": True,
                "login": user.get("login"),
                "name": user.get("name"),
                "user_id": user.get("id")
            }
        except Exception as e:
            return {"ok": False, "error": str(e)}

    def get_user(self):
        return self.request("GET", "/api/mine")

    def get_books(self):
        return self.request("GET", "/api/mine/books")

    def get_docs(self, book_id):
        return self.request("GET", f"/api/docs?book_id={book_id}")

    def get_doc(self, slug, book_id):
        return self.request("GET", f"/api/docs/{slug}?book_id={book_id}")

    def create_doc(self, book_id, title, body="", target_uuid="", doc_type="Doc"):
        """创建文档"""
        payload = {
            "action": "prependChild" if not target_uuid else "prependChild",
            "book_id": book_id,
            "title": title,
            "type": doc_type,
            "insert_to_catalog": True,
            "slug": "",
            "status": 0,
            "body_draft_asl": None
        }
        if target_uuid:
            payload["target_uuid"] = target_uuid
        return self.request("POST", "/api/docs", json=payload)

    def update_doc(self, doc_id, title=None, body=None):
        """更新文档(支持 markdown body)"""
        payload = {}
        if title is not None:
            payload["title"] = title
        if body is not None:
            payload["body"] = body
        return self.request("PUT", f"/api/docs/{doc_id}", json=payload)

    def append_doc_body(self, doc_id, new_content):
        """读取当前文档内容,追加 markdown 文本"""
        doc = self.get_doc_by_id(doc_id)
        current_body = doc.get("body", "") or ""
        updated_body = current_body + "\n\n" + new_content if current_body else new_content
        return self.update_doc(doc_id, body=updated_body)

    def get_doc_by_id(self, doc_id):
        """通过 ID 获取文档(需要先缓存 slug 和 book_id)"""
        if hasattr(self, "_doc_cache") and doc_id in self._doc_cache:
            slug = self._doc_cache[doc_id]["slug"]
            book_id = self._doc_cache[doc_id]["book_id"]
            return self.get_doc(slug, book_id).get("data", {})
        return {}

    def delete_doc(self, doc_id):
        return self.request("DELETE", f"/api/docs/{doc_id}")

    def search(self, query, search_type="content", tab="public", limit=10, page=1):
        """搜索"""
        params = {
            "q": query,
            "type": search_type,  # content, book, user
            "tab": tab,
            "limit": limit,
            "p": page
        }
        return self.request("GET", "/api/zsearch", params=params)

    def lock_doc(self, doc_id, uuid_str):
        return self.request("PUT", f"/api/docs/{doc_id}/lock", json={"uuid": uuid_str})


if __name__ == "__main__":
    client = YuqueClient()
    health = client.health_check()
    print(f"[{'OK' if health['ok'] else 'FAIL'}] YuqueClient initialized")
    if health["ok"]:
        print(f"       Login: {health['login']}")
        print(f"       Name: {health['name']}")
        print(f"       User ID: {health['user_id']}")
    else:
        print(f"       Error: {health.get('error', 'Unknown')}")
